stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift

```diff
@@ -64,12 +67,74 @@ func posixPipe() -> (readFD: CInt, writeFD: CInt) {
 /// stderr.
 public func spawnChild(_ args: [String])
   -> (pid: pid_t, stdinFD: CInt, stdoutFD: CInt, stderrFD: CInt) {
+  // The stdout, stdin, and stderr from the child process will be redirected
+  // to these pipes.
+  let childStdout = posixPipe()
+  let childStdin = posixPipe()
+  let childStderr = posixPipe()
+
+#if os(Android)
+  // posix_spawn isn't available on Android. Instead, we fork and exec.
+  // To correctly communicate the exit status of the child process to this
+  // (parent) process, we'll use this pipe.
+  let childToParentPipe = posixPipe()
+
+  let pid = fork()
+  precondition(pid >= 0, "fork() failed")
+  if pid == 0 {
+    // pid of 0 means we are now in the child process.
+    // Capture the output before executing the program.
+    dup2(childStdout.writeFD, STDOUT_FILENO)
+    dup2(childStdin.readFD, STDIN_FILENO)
+    dup2(childStderr.writeFD, STDERR_FILENO)
+
+    // Set the "close on exec" flag on the parent write pipe. This will
+    // close the pipe if the execve() below successfully executes a child
+    // process.
+    let closeResult = fcntl(childToParentPipe.writeFD, F_SETFD, FD_CLOEXEC)
+    let closeErrno = errno
+    precondition(
+      closeResult == 0,
+      "Could not set the close behavior of the child-to-parent pipe; " +
+      "errno: \(closeErrno)")
+
+    // Start the executable. If execve() does not encounter an error, the
+    // code after this block will never be executed, and the parent write pipe
+    // will be closed.
+    withArrayOfCStrings([Process.arguments[0]] + args) {
+      execve(Process.arguments[0], $0, _getEnviron())
+    }
+
+    // If execve() encountered an error, we write the errno encountered to the
+    // parent write pipe.
+    let errnoSize = sizeof(errno.dynamicType)
+    var execveErrno = errno
+    let writtenBytes = withUnsafePointer(&execveErrno) {
+      write(childToParentPipe.writeFD, UnsafePointer($0), errnoSize)
+    }
+
+    let writeErrno = errno
+    if writtenBytes > 0 && writtenBytes < errnoSize {
+      // We were able to write some of our error, but not all of it.
+      // FIXME: Retry in this case.
+      preconditionFailure("Unable to write entire error to child-to-parent " +
+                          "pipe.")
+    } else if writtenBytes == 0 {
+      preconditionFailure("Unable to write error to child-to-parent pipe.")
+    } else if writtenBytes < 0 {
+      preconditionFailure("An error occurred when writing error to " +
+                          "child-to-parent pipe; errno: \(writeErrno)")
+    }
+
+    // Close the pipe when we're done writing the error.
+    close(childToParentPipe.writeFD)
```

Comment:
OK, I have [an implementation that reads from the child process](https://gist.github.com/modocache/c035decc101f7d53be65c45b6cbdfae2). However, this implementation fails for any StdlibUnittest that expects a crash. I think the reason why is because of how `execve` behaves when the program it runs crashes, but I'm not sure.

[Documentation](http://man7.org/linux/man-pages/man2/execve.2.html) states that it "on success, `execve()` does not return". But in cases where a StdlibUnittest subprocess fails, I think it does return. This means that it writes `errno` to the parent pipe. When we read the `errno` out of the pipe, I don't think we have a way to tell whether the error has to do with `execve()` being unable to run a child process, or whether the child process crashed.

I might be missing something here, so feedback would be appreciated. Also, I still think we could hold off on this work for now. Once this pull request is merged, I can work on this more as part of #1714.